### ISSUE TITLE
No trivial way to disable default light when scene has no lights

### ISSUE THREAD

**Issue by el-dee** [panda3d/panda3d#1486]
## Description

If a  scene a no active light, Panda3D emulate the behaviour of the fixed pipeline of OpenGL and adds a fake light with color (1, 1, 1, 1)

In the forum it is said that to disable this fake light and have a scene with actually no lights one has to add an AmbientLight with color set to (0, 0, 0, 0) or any small value

However it is no longer the case, even when configuring that ambient light there is still that fake light configured. So far the only way I found to disable it is to always add a fake directional light that has (0, 0, 0, 0) as color.

I would say that this test should be completed with the number of ambient light and the default ihgt only enabled if the count is also 0 : 

https://github.com/panda3d/panda3d/blob/53f32e440e3d21ff8148119770b9016a64f4d1b8/panda/src/display/graphicsStateGuardian.cxx#L1573

## Steps to Reproduce

Here is a small example to reproduce the issue, it shows that the diffuse component of the first light contains (1, 1, 1, 1)

```python
from panda3d.core import  CardMaker, Shader, AmbientLight
from direct.showbase.ShowBase import ShowBase


def shader():
    return Shader.make(Shader.SL_GLSL,
                       vertex="""
#version 410

uniform mat4 p3d_ProjectionMatrix;
uniform mat4 p3d_ModelViewMatrix;

in vec4 p3d_Vertex;

void main() {
    gl_Position = p3d_ProjectionMatrix * (p3d_ModelViewMatrix * p3d_Vertex);
}
""",
                       fragment="""
#version 410

uniform struct p3d_LightSourceParameters {
    vec4 position;
    vec4 diffuse;
    vec4 specular;
    vec3 attenuation;
    vec3 spotDirection;
    float spotCosCutoff;
} p3d_LightSource[8];

out vec4 frag_color;

void main() {
    frag_color = p3d_LightSource[0].diffuse;
}
""")

base = ShowBase()
cm = CardMaker('card')
card = base.render.attach_new_node(cm.generate())
card.setPos(-0.5, 3, -0.5)
card.set_shader(shader())

ambient = AmbientLight("ambient")
ambient.set_color((0, 0, 0, 0))
ambient_np = base.render.attach_new_node(ambient)
base.render.set_light(ambient_np)

base.render.ls()

base.run()
```

## Environment
 * Operating system: Linux
 * System architecture: x86-64
 * Panda3D version: 1.10.x / 1.11.x
 * Installation method: SDK / From source
 * Python version (if using Python): 3.7 / 2.8
 * Compiler (if using C++): N/A

**Comment by serkkz** [comment]
Not sure why you are using `p3d_LightSource`. However, the required information contains `p3d_LightModel`


```python
from panda3d.core import  CardMaker, Shader, AmbientLight
from direct.showbase.ShowBase import ShowBase


def shader():
    return Shader.make(Shader.SL_GLSL,
                       vertex="""
#version 410

uniform mat4 p3d_ProjectionMatrix;
uniform mat4 p3d_ModelViewMatrix;

in vec4 p3d_Vertex;

void main() {
    gl_Position = p3d_ProjectionMatrix * (p3d_ModelViewMatrix * p3d_Vertex);
}
""",
                       fragment="""
#version 410

uniform struct p3d_LightModelParameters {
    vec4 ambient;
} p3d_LightModel;

out vec4 frag_color;

void main() {
    frag_color = p3d_LightModel.ambient;
}
""")

base = ShowBase()
cm = CardMaker('card')
card = base.render.attach_new_node(cm.generate())
card.setPos(-0.5, 3, -0.5)
card.set_shader(shader())

ambient = AmbientLight("ambient")
ambient.set_color((0, 0, 0, 0))
ambient_np = base.render.attach_new_node(ambient)
base.render.set_light(ambient_np)

base.render.ls()

base.run()
```

**Comment by serkkz** [comment]
Hmm, isn't that how it should be calculated.
```frag_color = p3d_LightModel.ambient * p3d_LightSource[0].diffuse;```
It's not entirely clear what the bug is.

By the way, everything works as expected.
```python
from direct.showbase.ShowBase import ShowBase
from panda3d.core import NodePath, AmbientLight

class MyApp(ShowBase):

    def __init__(self):
        ShowBase.__init__(self)

        scene = self.loader.load_model("models/environment")
        scene.reparent_to(render)
        scene.set_scale(0.25, 0.25, 0.25)
        scene.set_pos(-8, 42, 0)

        ambient = AmbientLight("ambient")
        ambient.set_color((0.2, 0.2, 0.2, 1))

        ambient_np = NodePath(ambient)
        ambient_np.reparent_to(render)

        render.set_light(ambient_np)
        #render.set_shader_auto()

app = MyApp()
app.run()
```

It bug is not an bug at all.

**Comment by el-dee** [comment]
@serkkz : Just to clarify, the problem arises only when using custom shaders. When using the auto shader or the fixed pipeline, setting a dummy ambient light is enough to disable the default light. But for the custom shaders, Panda3D still creates a default light even when a dummy ambient light is created. If you run the example above, you see that the card is fully white, but if you use the auto shader as in your own example, the card will be black as expected.

**Comment by serkkz** [comment]
Actually, I understood your problem. However, note that `vec4(1, 1, 1, 1)` by diffuse default makes sense when there are no light sources in the scene, does not allow you to broken the final lighting calculation.

Let's say you don't have any lighting in scene. There is only ambient lighting.

ambient                   diffuse
vec4(0.5, 0.5, 0.5, 0.5)*vec4(0, 0, 0, 0) = calculation is broken

ambient                   diffuse
vec4(0.5, 0.5, 0.5, 0.5)*vec4(1, 1, 1, 1) = result is expected

**Comment by rdb** [comment]
I am 90% sure this is a bug, but we need to compare against the behaviour of the FFP to be sure.  If any lights are present we are supposed to enable lighting, whether that's an ambient light or not.

That said, maybe we need to add an explicit lighting enable flag to LightAttrib to make it more explicit.

**Comment by serkkz** [comment]
@el-dee 
The problem arises actually from the fact that you ignore `p3d_LightModel.ambient` in your calculations.

**Comment by el-dee** [comment]
> That said, maybe we need to add an explicit lighting enable flag to LightAttrib to make it more explicit.

+1 for the enable flag, it would be much cleaner that way that construct a fake light (whatever it is) to disable that behaviour. I was thinking of a PRC flag, but on the LightAttrib it would be fine too

**Comment by el-dee** [comment]
> @el-dee The problem arises actually from the fact that you ignore `p3d_LightModel.ambient` in your calculations.

I think we are misunderstanding each other: I'm not doing any lighting calculation in the example; it's just to show the content of the diffuse color of the first light source. And it's the color of the light, not the diffuse color of a material so I can not in any case multiply it with the global ambient light color, it has no sense as both lights are independents.

**Comment by serkkz** [comment]
The general rule for calculating illumination will be:
`ambient*diffuse`
Question what will `p3d_LightSource` provide if there are no light sources? 
This vector is just a stub.

**Comment by rdb** [comment]
This is not correct.  The ambient value is *added*, not *multiplied* with the light summation.

The alpha issue is interesting, it may be that the alpha value needs to be clamped from 0 to 1.

**Comment by serkkz** [comment]
I think the artifact with alpha is the result of the intervention of `vec4(1, 1, 1, 1)`, that is, the conclusions of el-dee are correct about the bug. At the moment it is fixed like this.

```glsl
total_light += vec4(p3d_LightSource[i].diffuse.xyz, 0);
```
I commented on the directional light, and now I get the stage lighting brightly. This is definitely a mistake, I was misled due to my lack of experience using textures with an alpha channel.

**Comment by ZzzhHe** [comment]
Hi, has this bug been fixed? I followed the provided example to reproduce the issue, but instead of the card appearing white, it rendered black. Any updates or additional insights on this?
